File: LinkDemo-ObjC/LinkDemo-ObjC/AppDelegate+Preflight.m
//
//  AppDelegate+Preflight.m
//  LinkDemo-ObjC
//
//  Copyright © 2020 Plaid Inc. All rights reserved.
//

#import <LinkKit/LinkKit.h>

#import "AppDelegate+Preflight.h"

@implementation AppDelegate (Preflight)

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
    // NOTE: Below there are Plaid setup examples for a preflight check, yet with recent changes to
    // Plaid's API and LinkKit this is no longer a necessity.
//    [self setupPlaidLinkWithCustomConfiguration];
//    [self setupPlaidLinkWithSharedConfiguration];
    return YES;
}

// MARK: Plaid Link setup with shared configuration from Info.plist
- (void)setupPlaidLinkWithSharedConfiguration {

    // <!-- SMARTDOWN_SETUP_SHARED -->
    // With shared configuration from Info.plist
    [PLKPlaidLink setupWithSharedConfiguration:^(BOOL success, NSError * _Nullable error) {
        if (success) {
            // Handle success here, e.g. by posting a notification
            NSLog(@"Plaid Link setup was successful");
            [[NSNotificationCenter defaultCenter] postNotificationName:@"PLDPlaidLinkSetupFinished" object:self];
        }
        else {
            NSLog(@"Unable to setup Plaid Link due to: %@", [error localizedDescription]);
        }
    }];
    // <!-- SMARTDOWN_SETUP_SHARED -->

}

// MARK: Plaid Link setup with custom configuration
- (void)setupPlaidLinkWithCustomConfiguration {

    #warning Replace <#YOUR_PLAID_PUBLIC_KEY#> below with your public_key
    // <!-- SMARTDOWN_SETUP_CUSTOM -->
    // With custom configuration
    PLKConfiguration* linkConfiguration;
    @try {
        linkConfiguration = [[PLKConfiguration alloc] initWithKey:@"<#YOUR_PLAID_PUBLIC_KEY#>"
                                                              env:PLKEnvironmentDevelopment
                                                          product:PLKProductAuth];
        linkConfiguration.clientName = @"Link Demo";
        [PLKPlaidLink setupWithConfiguration:linkConfiguration completion:^(BOOL success, NSError * _Nullable error) {
            if (success) {
                // Handle success here, e.g. by posting a notification
                NSLog(@"Plaid Link setup was successful");
                [[NSNotificationCenter defaultCenter] postNotificationName:@"PLDPlaidLinkSetupFinished" object:self];
            }
            else {
                NSLog(@"Unable to setup Plaid Link due to: %@", [error localizedDescription]);
            }
        }];
    } @catch (NSException *exception) {
        NSLog(@"Invalid configuration: %@", exception);
    }
    // <!-- SMARTDOWN_SETUP_CUSTOM -->

}

@end

File: LinkDemo-Swift/LinkDemo-Swift/AppDelegate+Preflight.swift
//
//  AppDelegate+Preflight.swift
//  LinkDemo-Swift
//
//  Copyright © 2020 Plaid Inc. All rights reserved.
//

import LinkKit

extension AppDelegate {

    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
        // NOTE: Below there are Plaid setup examples for a preflight check, yet with recent changes to
        // Plaid's API and LinkKit this is no longer a necessity.
//        setupPlaidWithCustomConfiguration()
//        setupPlaidLinkWithSharedConfiguration()
        return true
    }

    // MARK: Plaid Link setup with shared configuration from Info.plist
    func setupPlaidLinkWithSharedConfiguration() {

        // <!-- SMARTDOWN_SETUP_SHARED -->
        // With shared configuration from Info.plist
        PLKPlaidLink.setup { (success, error) in
            if (success) {
                // Handle success here, e.g. by posting a notification
                NSLog("Plaid Link setup was successful")
                NotificationCenter.default.post(name: NSNotification.Name(rawValue: "PLDPlaidLinkSetupFinished"), object: self)
            }
            else if let error = error {
                NSLog("Unable to setup Plaid Link due to: \(error.localizedDescription)")
            }
            else {
                NSLog("Unable to setup Plaid Link")
            }
        }
        // <!-- SMARTDOWN_SETUP_SHARED -->

    }

    // MARK: Plaid Link setup with custom configuration
    func setupPlaidWithCustomConfiguration() {

        #warning("Replace <#YOUR_PLAID_PUBLIC_KEY#> below with your public_key")
        // <!-- SMARTDOWN_SETUP_CUSTOM -->
        // With custom configuration
        let linkConfiguration = PLKConfiguration(key: "<#YOUR_PLAID_PUBLIC_KEY#>", env: .development, product: .auth)
        linkConfiguration.clientName = "Link Demo"
        PLKPlaidLink.setup(with: linkConfiguration) { (success, error) in
            if (success) {
                // Handle success here, e.g. by posting a notification
                NSLog("Plaid Link setup was successful")
                NotificationCenter.default.post(name: NSNotification.Name(rawValue: "PLDPlaidLinkSetupFinished"), object: self)
            }
            else if let error = error {
                NSLog("Unable to setup Plaid Link due to: \(error.localizedDescription)")
            }
            else {
                NSLog("Unable to setup Plaid Link")
            }
        }
        // <!-- SMARTDOWN_SETUP_CUSTOM -->

    }

}
